control-plane/agents/src/bin/core/volume/service.rs

```diff
@@ -578,6 +580,29 @@ impl Service {
     #[tracing::instrument(level = "info", skip(self), err, fields(volume.uuid = %request.uuid))]
     pub(super) async fn resize_volume(&self, request: &ResizeVolume) -> Result<Volume, SvcError> {
         let mut volume = self.specs().volume(&request.uuid).await?;
-        volume.resize(&self.registry, request).await
+        let required = request.requested_size - volume.as_ref().size;
+
+        // If there is a defined system wide capacity limit, ensure we don't breach that.
+        if request.cluster_capacity_limit().is_some() {
+            let mut current_borrowed = self.capacity_limit_borrow.write();
+            *current_borrowed += required;
+            self
+                .specs()
+                .check_capacity_limit_for_resize(request, *current_borrowed).map_err(|e| {
+                    tracing::error!(volume=%volume.uuid(),"resize_volume: return borrowed capacity limit {required}B, current {}B ", *current_borrowed);
+                    *current_borrowed -= required;
+                    e
+                })?;
+        }
+
+        let resize_ret = volume.resize(&self.registry, request).await;
+        if request.cluster_capacity_limit().is_some() {
+            // Reset the capacity limit that we consumed and will now be accounted in the system's
+            // current total.
+            let mut current_borrowed = self.capacity_limit_borrow.write();
+            tracing::info!(volume=%volume.uuid(),"resize_volume: return borrowed capacity limit {required}B, current {}B ", *current_borrowed);
+            *current_borrowed -= required;
+        }
+        resize_ret
```

Comment:
Thanks, looks cleaner. Except that I had to read and get the value before calling checks, otherwise the guard doesn't get dropped, and the write() call hangs in `map_err` path.
```
        let current = *self.capacity_limit_borrow.read()
        self.specs()
                .check_capacity_limit_for_resize(limit, current)
```